velox/expression/tests/ArrayProxyTest.cpp

```diff
@@ -0,0 +1,253 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glog/logging.h"
+#include "gtest/gtest.h"
+#include "velox/expression/VectorUdfTypeSystem.h"
+#include "velox/functions/Udf.h"
+#include "velox/functions/prestosql/tests/FunctionBaseTest.h"
+
+namespace facebook::velox {
+namespace {
+// Function that creates array with values 0...n-1.
+// Uses all possible functions in the array proxy interface.
+template <typename T>
+struct Func {
+  template <typename TOut>
+  bool call(TOut& out, const int64_t& n) {
+    for (int i = 0; i < n; i++) {
+      switch (i % 4) {
+        case 0:
+          out.add_item() = i;
+          break;
+        case 1:
+          out.push_back(i);
+          break;
+        case 2:
+          out.add_null();
+          break;
+        case 3:
+          out.resize(out.size() + 1);
+          out[out.size() - 1] = i;
+          break;
+      }
+    }
+    return true;
+  }
+};
+
+class ArrayProxyTest : public functions::test::FunctionBaseTest {
+ public:
+  VectorPtr prepareResult(
+      const std::shared_ptr<Type>& arrayType,
+      vector_size_t size_ = 1) {
+    VectorPtr result;
+    BaseVector::ensureWritable(
+        SelectivityVector(size_), arrayType, this->execCtx_.pool(), &result);
+    return result;
+  }
+
+  template <typename T>
+  void testE2E() {
+    registerFunction<Func, ArrayProxyT<T>, int64_t>({"func"});
+
+    auto result = evaluate(
+        "func(c0)",
+        makeRowVector(
+            {makeFlatVector<int64_t>({1, 2, 3, 4, 5, 6, 7, 8, 9, 10})}));
+
+    std::vector<std::vector<std::optional<T>>> expected;
+    for (auto i = 1; i <= 10; i++) {
+      expected.push_back({});
+      auto& currentExpected = expected[expected.size() - 1];
+      for (auto j = 0; j < i; j++) {
+        switch (j % 4) {
+          case 0:
+            currentExpected.push_back(j);
+            break;
+          case 1:
+            currentExpected.push_back(j);
+            break;
+          case 2:
+            currentExpected.push_back(std::nullopt);
+            break;
+          case 3:
+            currentExpected.push_back(j);
+            break;
+        }
+      }
+    }
+    assertEqualVectors(result, makeNullableArrayVector(expected));
+  };
+
+  struct TestWriter {
+    VectorPtr result_;
+    exec::VectorWriter<ArrayProxyT<int64_t>> writer_;
+  };
+
+  std::shared_ptr<TestWriter> makeTestWriter() {
+    std::shared_ptr<TestWriter> writer;
+
+    writer->result_ =
+        prepareResult(std::make_shared<ArrayType>(ArrayType(BIGINT())));
+    writer->writer_.init(*writer->result_.get()->as<ArrayVector>());
+    writer->writer_.setOffset(0);
+    return writer;
+  }
+};
+
+TEST_F(ArrayProxyTest, ArrayAddNull) {
```

Comment:
naming: lowercase test names, e.g. arrayAddNull. If all tests names have "array" prefix, then I'd drop it, e.g. `TEST_F(ArrayProxyTest, addNull)`